qiskit_optimization/converters/linear_inequality_to_penalty.py

```diff
@@ -0,0 +1,335 @@
+            elif x.vartype == Variable.Type.BINARY:
+                self._dst.binary_var(x.name)
+            elif x.vartype == Variable.Type.INTEGER:
+                self._dst.integer_var(x.lowerbound, x.upperbound, x.name)
+            else:
+                raise QiskitOptimizationError("Unsupported vartype: {}".format(x.vartype))
+
+        # get original objective terms
+        offset = problem.objective.constant
+        linear = problem.objective.linear.to_dict()
+        quadratic = problem.objective.quadratic.to_dict()
+        sense = problem.objective.sense.value
+
+        # convert linear constraints into penalty terms
+        for constraint in problem.linear_constraints:
+
+            # special constraint check function here
+            if not self._is_special_constraint(problem, constraint):
+                self._dst.linear_constraints.append(constraint)
+                continue
+            #
+
+            conv_matrix = self._conversion_matrix(constraint)
+            rowlist = list(constraint.linear.to_dict().items())
+
+            # constant part
+            if conv_matrix[0][0] != 0:
+                offset += sense * penalty * conv_matrix[0][0]
+
+            # linear parts of penalty
+            for (j, z) in enumerate(rowlist):
+                # if j already exists in the linear terms dic, add a penalty term
+                # into existing value else create new key and value in the linear_term dict
+
+                if conv_matrix[0][j + 1] != 0:
+                    linear[z[0]] = linear.get(z[0], 0.0) + sense * penalty * conv_matrix[0][j + 1]
+
+            # quadratic parts of penalty
+            for (j, z) in enumerate(rowlist):
+                for k in range(j, len(rowlist)):
+                    # if j and k already exist in the quadratic terms dict,
+                    # add a penalty term into existing value
+                    # else create new key and value in the quadratic term dict
+
+                    if conv_matrix[j + 1][k + 1] != 0:
+                        tup = cast(Union[Tuple[int, int], Tuple[str, str]], (z[0], rowlist[k][0]))
+                        quadratic[tup] = (
+                            quadratic.get(tup, 0.0) + sense * penalty * conv_matrix[j + 1][k + 1]
+                        )
+
+        # Copy quadratic_constraints
+        for quadratic_constraint in problem.quadratic_constraints:
+            self._dst.quadratic_constraints.append(quadratic_constraint)
+
+        if problem.objective.sense == QuadraticObjective.Sense.MINIMIZE:
+            self._dst.minimize(offset, linear, quadratic)
+        else:
+            self._dst.maximize(offset, linear, quadratic)
+
+        # Update the penalty to the one just used
+        self._penalty = penalty  # type: float
+
+        return self._dst
+
+    def _conversion_matrix(self, constraint) -> np.ndarray:
+        """Construct conversion matrix for special constraint.
+
```

Comment:
Could you write how the conversion matrix works briefly, for example, what `conv_matrix[i][j]` corresponds to.

Reply:
I would recommend to return `offset`, `linear`, and `quadratic` separately instead of `conv_matrix`, which combines `offset`, `linear`, and `quadratic` all because it looks tricky (you need to do +1 or -1 of variables index).

Reply:
conv_matrix[0][0] is offset.
conv_matrix[0][i] (i >= 1) is linear
conv_matrix[i][j] (i>=1, j>=1 i < j) is quadratic
so this matrix is a combination of offset, linear, and quadratic in a single matrix so to speak. I'd like to postpone suggested change in the future.